File: Leetcode/dfs_recursion_backtracking/473_MatchsticksToSquare/Solution.java
class Solution {
    public boolean makesquare(int[] nums) {
        Arrays.sort(nums);
        int sum = 0;
        for(int num: nums){
            sum += num;
        }
        if( sum==0 || sum%4!=0 )
            return false;
        int[] edges = {sum/4, sum/4, sum/4, sum/4};
        return dfs(nums, nums.length-1, edges);
    }
    private boolean dfs(int[] nums, int index, int[] edges){
        if(index<0)
            return true;
        for(int i = 0; i<4; ++i){
            if(edges[i] >= nums[index]){
                edges[i] -= nums[index];
                if(dfs(nums, index-1, edges)){
                    return true;
                }
                edges[i] += nums[index];
            }
        }
        return false;
    }
}

File: Leetcode/dfs_recursion_backtracking/473_MatchsticksToSquare/Solution.py
class Solution(object):
    def makesquare(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        nums.sort(reverse = True)
        numSum = sum(nums)
        if numSum == 0 or numSum%4 != 0:
            return False
        def dfs(nums, edges):
            if not nums:
                return True
            for i in range(4):
                if edges[i] >= nums[0]:
                    edges[i] -= nums[0]
                    if dfs(nums[1:], edges):
                        return True
                    edges[i] += nums[0]
            return False
        return dfs(nums, [numSum//4 for i in range(4)])
